controllers/machine_webhook.go

```diff
@@ -0,0 +1,101 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/pkg/errors"
+	"k8s.io/api/admission/v1beta1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	infrav1 "sigs.k8s.io/cluster-api-provider-aws/api/v1alpha3"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	"sigs.k8s.io/cluster-api/controllers/external"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+// +kubebuilder:webhook:path=/validate-infrastructure-cluster-x-k8s-io-v1alpha3-machine,mutating=false,failurePolicy=fail,groups=cluster.x-k8s.io,resources=machines,verbs=create,versions=v1alpha3,name=providervalidation.machine.infrastructure.cluster.x-k8s.io
+
+// MachineWebhook handles Machine admission webhook requests.
```

Comment:
This webhook looks quite different from the other webhooks we have around.

I'm wondering if there is a way to use the nice building blocks offered by kubebuilder by using a type alias like
```golang
// +kubebuilder:object:root=true
type CoreMachine struct {
	clusterv1.Machine
}
```

and have the registration happen as usual with:

```golang
func (r *CoreMachine) SetupWebhookWithManager(mgr ctrl.Manager) error {
	return ctrl.NewWebhookManagedBy(mgr).
		For(r).
		Complete()
}
```

Reply:
Hi Vince!  I'm 100% in agreement and started the implementation using the method above. The biggest challenge is that the Validating webhook methods don't allow for client injection and we need to use the client to retrieve the AWSMachine to see if it has `spec.ami.id` set. If so, we can approve a Machine without a `spec.version`

Here are some related controller runtime issues:
https://github.com/kubernetes-sigs/controller-runtime/issues/711
https://github.com/kubernetes-sigs/controller-runtime/issues/451

Reply:
This is intriguing. I'm not sure about using a client to fetch other objects in webhook validation. There's two side effects:
a) Someone doing a `kubectl apply -f` would need to order their yaml files, and that's weird, and affects the ability to use third-party tools to apply manifests.
b) Performance impacts. Validation webhooks should be for fairly lightweight checks.

May be better to defer checking the version to the controller, and just error the AWSMachine created to fix the NPE.

Reply:
Yeah, we've gone back and forth whether a webhook is the right solution here. Tried to solve it completely using webhooks and it's not possible.

So the question is whether we want a Webhook to catch the issue when possible. No one needs to order their YAML files ... if the order is a certain way, the Webhook can catch it. If not it fails.

Not sure of the performance impact.  I've posed it as an outstanding question (earlier in the PR comments) if we want a webhook or if we should just let the controller catch it.  I'm good with either ... just looking for feedback. Thanks for reviewing.

Reply:
It should be reading from the cache instead of doing a live lookup, so I wouldn't expect any network latency to impact things here.

However, to keep things simple, how about we just merge the code you have that fixes the NPE? You can hold on to a copy of the webhook code and we can reevaluate it later. WDYT?

Reply:
Absolutely!  since the webhook can't really prevent this in all cases, maybe it's just not the right solution